File: c++/16637.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> operands;
vector<char> operators;
int ans = -1e9;

int calc(char oper, int a, int b) {
  if (oper == '+')
    return a + b;
  else if (oper == '-')
    return a - b;
  else
    return a * b;
}

void dfs(int idx, int result) {
  if (idx == operators.size()) {
    ans = max(ans, result);
    return;
  }

  dfs(idx + 1, calc(operators[idx], result, operands[idx + 1]));

  if (idx + 2 <= operators.size()) {
    int temp = calc(operators[idx + 1], operands[idx + 1], operands[idx + 2]);
    dfs(idx + 2, calc(operators[idx], result, temp));
  }
}

int main() {
  freopen("example.txt", "r", stdin);
  int N;
  cin >> N;

  string s;
  cin >> s;

  for (int i = 0; i < N; i++) {
    if (i % 2 == 0)
      operands.push_back(s[i] - '0');
    else
      operators.push_back(s[i]);
  }

  dfs(0, operands[0]);

  cout << ans;

  return 0;
}

File: js/16637.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const N = +inputs[0];
const s = inputs[1].split('');
const operators = s.filter((_, idx) => idx % 2 === 1);
const operands = s.filter((_, idx) => idx % 2 === 0).map(Number);

const n = Math.floor(N / 2);

const arr = [];
const isVisited = Array(n).fill(false);
let result = -Infinity;

dfs(0);
console.log(result);

function dfs(idx) {
  calc();

  for (let i = idx; i < n; i++) {
    if (isVisited[i]) continue;
    isVisited[i] = true;
    arr.push(i);
    dfs(i + 2);
    arr.pop();
    isVisited[i] = false;
  }
}

function calc() {
  const copyOperators = [...operators];
  const copyOperands = [...operands];

  for (let i = arr.length - 1; i >= 0; i--) {
    const loc = arr[i];
    const operator = copyOperators[loc];

    let res;
    if (operator === '+') {
      res = copyOperands[loc] + copyOperands[loc + 1];
    } else if (operator === '-') {
      res = copyOperands[loc] - copyOperands[loc + 1];
    } else if (operator === '*') {
      res = copyOperands[loc] * copyOperands[loc + 1];
    }

    copyOperators.splice(loc, 1);
    copyOperands.splice(loc, 2, res);
  }

  let k = copyOperands[0];

  for (let i = 0; i < copyOperators.length; i++) {
    const operator = copyOperators[i];

    if (operator === '+') {
      k += copyOperands[i + 1];
    } else if (operator === '-') {
      k -= copyOperands[i + 1];
    } else if (operator === '*') {
      k *= copyOperands[i + 1];
    }
  }

  result = Math.max(k, result);
}
